### PR TITLE
Add support for migrating external location permissions from interactive cluster mounts

### LINKED ISSUE #1192: [FEATURE]: Create External Location ACL for Principals based on Interactive cluster permission using SPN
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

Currently Users access interactive clusters using SPN configured which has permission on the underlying azure storage.
When migrating to UC, we need to create permission to interactive clusters users on the external location, catalog and schema  based on the underlying spn access on the storage

Related issues:

- https://github.com/databrickslabs/ucx/issues/907

### Proposed Solution

 - identify interactive clusters with spn configured.
 - check spn permission on the underlying storage
 - grant appropriate permission on the external location to the users of the interactive clusters
 - grant use permission on the catalog and schema equivalent tin UC for the users of the interactive clusters

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by HariGS-DB**
## Changes
 - Gets the locations of interactive cluster from the AzureACL class
 - Gets the interactive cluster user mapping
 - for each location under each cluster grants CREATE EXTERNAL TABLE, CREATE EXTERNAL VOLUME and READ_FILES permission to each cluster principal

Linked issues #1192 #1193 
Resolves #1192 #1193

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [X] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests

- [X] manually tested
- [X] added unit tests
- [X] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

File: src/databricks/labs/ucx/azure/locations.py [hunk 1]
```diff
@@ -186,3 +193,42 @@ def run(self):
 
         logger.info("All UC external location are created.")
         return leftover_loc_urls
+
+    def _apply_location_acl(
+        self,
```
Comment (nfx) [inline src/databricks/labs/ucx/azure/locations.py:198]:
```suggestion
        self
```

File: src/databricks/labs/ucx/azure/locations.py [hunk 2]
```diff
@@ -186,3 +193,42 @@ def run(self):
...
+        # get the eligible location mapped for each interactive cluster
+        cluster_locations = self._azure_acl.get_eligible_locations_principals()
+        for cluster_id, locations in cluster_locations.items():
+            # get interactive cluster users
+            principals = self._principal_acl.get_cluster_principal_mapping(cluster_id)
+            if len(principals) == 0:
+                continue
+            for location_url, _ in locations.items():
+                # get the location name for the given url
+                location_name = self._get_location_name(location_url)
```
Comment (nfx) [inline src/databricks/labs/ucx/azure/locations.py:217]: Bug: skip location if none and log a warning

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 3]
```diff
@@ -612,3 +625,43 @@ def _get_cluster_principal_mapping(self, cluster_id: str) -> list[str]:
             if acl.service_principal_name is not None:
                 principal_list.append(acl.service_principal_name)
         return principal_list
+
+    def apply_location_acl(
+        self,
+    ):
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/grants.py:631]:
```suggestion
    def apply_location_acl(self):
```

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 4]
```diff
@@ -341,6 +354,8 @@ def grants(
 
 
 class AwsACL:
+    INSTANCE_PROFILES_FILE_NAMES: typing.ClassVar[str] = "aws_instance_profile_info.csv"
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/grants.py:357]:
```suggestion
    INSTANCE_PROFILES_FILE_NAME: typing.ClassVar[str] = "aws_instance_profile_info.csv"
```
Has to be singular

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 5]
```diff
@@ -390,9 +405,7 @@ def get_eligible_locations_principals(self) -> dict[str, dict]:
             logger.error(msg)
             raise ResourceDoesNotExist(msg) from None
 
-        permission_mappings = self._installation.load(
-            list[AWSRoleAction], filename=AWSResourcePermissions.INSTANCE_PROFILES_FILE_NAMES
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/grants.py:394]: Why did you remove this? Is this file managed by AwsACL? What if someone renames the filename in resource permissions class?

Revert this removal.

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 6]
```diff
@@ -596,7 +609,7 @@ def _get_grants(
 
         return grants
 
-    def _get_cluster_principal_mapping(self, cluster_id: str) -> list[str]:
+    def get_cluster_principal_mapping(self, cluster_id: str) -> list[str]:
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/grants.py]: Why do we need this method public? We don't use it outside of this ass. Revert.

File: tests/integration/azure/test_locations.py [hunk 7]
```diff
@@ -185,11 +203,51 @@ def test_overlapping_location(caplog, ws, sql_backend, inventory_schema):
...
+    try:
+        location_migration.run()
+        permissions = ws.grants.get(
+            SecurableType.EXTERNAL_LOCATION, env_or_skip("TEST_A_LOCATION"), principal=user.user_name
+        )
+        expected_permission = PrivilegeAssignment(
+            principal=user.user_name,
+            privileges=[Privilege.CREATE_EXTERNAL_TABLE, Privilege.CREATE_EXTERNAL_VOLUME, Privilege.READ_FILES],
+        )
+        assert expected_permission in permissions.privilege_assignments
```
Comment (nfx) [inline tests/integration/azure/test_locations.py:242]: Does this work?...

File: tests/integration/conftest.py [hunk 8]
```diff
@@ -444,10 +445,76 @@ def azure_cli_authenticated(self):
...
+        return WorkspaceConfig(
+            warehouse_id=self._env_or_skip("TEST_DEFAULT_WAREHOUSE_ID"),
+            inventory_database=self.inventory_database,
+            connect=self.workspace_client.config,
+            renamed_group_prefix=f'tmp-{self.inventory_database}-',
+        )
+
+    @cached_property
+    def installation(self):
+        return MockInstallation()
```
Comment (nfx) [inline tests/integration/conftest.py:459]: Are you sure you want to use in-memory installation for something that calls a job?..

File: tests/integration/conftest.py [hunk 9]
```diff
@@ -428,9 +428,10 @@ def runtime_ctx(ws, sql_backend, make_table, make_schema, make_udf, make_group,
 
 
 class LocalAzureCliTest(WorkspaceContext):
```
Comment (nfx) [inline tests/integration/conftest.py:430]: What if you extend the TestRuntimeContext or TestInstallationContext, so that you don't have to copy-paste so many cached properties?..

File: tests/unit/aws/test_access.py [hunk 10]
```diff
@@ -121,7 +122,13 @@ def test_create_external_locations(mock_ws, installation_multiple_roles, mock_aw
...
     ]
     aws_resource_permissions = AWSResourcePermissions(
-        installation_multiple_roles, mock_ws, backend, mock_aws, locations, "ucx"
+        installation_multiple_roles,
+        mock_ws,
+        backend,
+        mock_aws,
+        locations,
+        "ucx",
+        create_autospec(PrincipalACL),
```
Comment (nfx) [inline tests/unit/aws/test_access.py:131]: Blocker: you've added mocks for AwsACL, but you don't assert if methods from those mocks called, making these tests broken

File: tests/integration/azure/test_locations.py [hunk 11]
```diff
@@ -20,7 +27,7 @@ def save_delete_location(ws, name):
 
 
 @pytest.mark.skip
-def test_run(caplog, ws, sql_backend, inventory_schema):
+def test_run(caplog, ws, sql_backend, inventory_schema, runtime_ctx):
```
Comment (nfx) [inline tests/integration/azure/test_locations.py:30]:
```suggestion
def test_run(caplog, ws, sql_backend, inventory_schema, azure_cli_ctx):
```
Rewrite to use the correct test context

File: tests/integration/azure/test_locations.py [hunk 12]
```diff
@@ -100,7 +110,11 @@ def test_read_only_location(caplog, ws, sql_backend, inventory_schema):
     azurerm = AzureResources(azure_mgmt_client, graph_client)
 
     location_migration = ExternalLocationsMigration(
-        ws, location_crawler, AzureResourcePermissions(installation, ws, azurerm, location_crawler), azurerm
+        ws,
+        location_crawler,
+        AzureResourcePermissions(installation, ws, azurerm, location_crawler),
+        azurerm,
```
Comment (nfx) [inline tests/integration/azure/test_locations.py:116]: Rewrite to make use of correct context

File: tests/integration/azure/test_locations.py [hunk 13]
```diff
@@ -150,7 +168,7 @@ def test_missing_credential(caplog, ws, sql_backend, inventory_schema):
 
 
 @pytest.mark.skip
-def test_overlapping_location(caplog, ws, sql_backend, inventory_schema):
+def test_overlapping_location(caplog, ws, sql_backend, inventory_schema, runtime_ctx):
```
Comment (nfx) [inline tests/integration/azure/test_locations.py:171]: Rewrite

File: tests/integration/aws/test_credentials.py [hunk 14]
```diff
@@ -13,7 +13,7 @@
 
 
 @pytest.fixture
-def run_migration(ws, sql_backend, env_or_skip):
+def run_migration(ws, sql_backend, env_or_skip, runtime_ctx):
```
Comment (nfx) [inline tests/integration/aws/test_credentials.py:16]:
```suggestion
def run_migration(ws, sql_backend, env_or_skip, aws_cli_ctx):
```
Create aws cli context that skips on the absence of AWS profile

File: tests/integration/aws/test_access.py [hunk 15]
```diff
@@ -62,7 +64,9 @@ def test_create_external_location(ws, env_or_skip, make_random, inventory_schema
     assert external_location[0].credential_name == f"ucx_{rand}"
 
 
-def test_create_uber_instance_profile(ws, env_or_skip, make_random, inventory_schema, sql_backend, make_cluster_policy):
+def test_create_uber_instance_profile(
```
Comment (nfx) [inline tests/integration/aws/test_access.py:67]: Rewrite to use aws cli context

**Review by nfx — CHANGES_REQUESTED** [review]
Also fix failing checks

File: src/databricks/labs/ucx/hive_metastore/grants.py (hunk 5 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/hive_metastore/grants.py:394]: I am calling PrincipalACL from AWSResourcePermission and this code is calling AWSResourcePermission and I am getting circular references

File: src/databricks/labs/ucx/azure/locations.py [hunk 16]
```diff
@@ -171,6 +174,7 @@ def run(self):
                 migrated_loc_urls.append(migrated_loc_url)
 
         leftover_loc_urls = [url for url in missing_loc_urls if url not in migrated_loc_urls]
+        self._principal_acl.apply_location_acl()
```
Comment (nkvuong) [inline src/databricks/labs/ucx/azure/locations.py:177]: on Azure, you're applying location acl during external location migration, but there is no circular dependencies, this is because Azure has a separate class for ext location migrations

**Review by nkvuong — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/hive_metastore/grants.py (hunk 5 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/hive_metastore/grants.py:394]: Good point

File: tests/integration/azure/test_locations.py (hunk 7 above)
Comment (HariGS-DB) [reply tests/integration/azure/test_locations.py:242]: @nfx I need access to azure labs mi access connector read only

File: src/databricks/labs/ucx/azure/locations.py (hunk 16 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/azure/locations.py:177]: https://github.com/databrickslabs/ucx/issues/1572 created issue to track this

File: src/databricks/labs/ucx/hive_metastore/grants.py (hunk 5 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/hive_metastore/grants.py:394]: https://github.com/databrickslabs/ucx/issues/1572 created issue to track this

**Comment by HariGS-DB** [comment]
Local test run for the int test test_create_external_location_validate_acl on aws
image

**Comment by nfx** [comment]
@HariGS-DB what about labs-azure-account-admin?

**Comment by HariGS-DB** [comment]
@nfx i get this error when i run my azure test. I need read access on the access-connector/MI on azure to complete my test.

'hari.selvarajan@databricks.com' with object id '998e34f7-45ce-406a-9cfc-2aff836d4c2b' does not have authorization to perform action 'Microsoft.Databricks/accessConnectors/read' over scope '/subscriptions/bf472657-510a-4fad-a050-87cfb4e1a2ce/resourceGroups/labs-azure-rg/providers/Microsoft.Databricks/accessConnectors/labs-azure-sandbox-mi' or the scope is invalid. If access was recently granted, please refresh your credentials."}

**Comment by HariGS-DB** [comment]
Azure Int test evidence
image

File: tests/integration/azure/test_locations.py [hunk 17]
```diff
@@ -20,7 +27,7 @@ def save_delete_location(ws, name):
 
 
 @pytest.mark.skip
```
Comment (nkvuong) [inline tests/integration/azure/test_locations.py:29]: btw, these can be replaced with `env_or_skip("IDE_PROJECT_ROOTS")`

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 18]
```diff
@@ -341,6 +348,9 @@ def grants(
 
 
 class AwsACL:
+    # adding this profile file name here to avoid circular references with aws/access.py
+    INSTANCE_PROFILES_FILE_NAME: typing.ClassVar[str] = "aws_instance_profile_info.csv"
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/grants.py:352]:
```suggestion
    FILENAME: typing.ClassVar[str] = "aws_instance_profile_info.csv"
```
rename

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 19]
```diff
@@ -612,3 +620,39 @@ def _get_cluster_principal_mapping(self, cluster_id: str) -> list[str]:
...
+            "CREATE EXTERNAL VOLUME and READ_FILES for existing eligible interactive cluster users"
+        )
+        # get the eligible location mapped for each interactive cluster
+        permissions = [Privilege.CREATE_EXTERNAL_TABLE, Privilege.CREATE_EXTERNAL_VOLUME, Privilege.READ_FILES]
+        for cluster_id, locations in self._cluster_locations.items():
+            # get interactive cluster users
+            principals = self._get_cluster_principal_mapping(cluster_id)
+            if len(principals) == 0:
+                continue
+            for location_url in locations.keys():
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/grants.py:641]: reduce nesting

File: tests/integration/__init__.py [hunk 20]
```diff
@@ -0,0 +1,12 @@
+def get_azure_spark_conf():
```
Comment (nfx) [inline tests/integration/__init__.py:1]: this must not be in this file, because everything in the integration testing must be almost real. this is dummy.

why it's not in conftest.py?

File: tests/integration/aws/test_access.py [hunk 21]
```diff
@@ -103,3 +108,43 @@ def test_create_uber_instance_profile(ws, env_or_skip, make_random, inventory_sc
 
     role_name = AWSInstanceProfile(instance_profile_arn).role_name
     aws.delete_instance_profile(role_name, role_name)
+
+
+def test_create_external_location_validate_acl(
+    make_cluster_permissions, ws, make_user, make_cluster, aws_cli_ctx, env_or_skip
```
Comment (nfx) [inline tests/integration/aws/test_access.py:114]:
```suggestion
    make_cluster_permissions, ws, make_user, make_cluster, aws_cli_ctx, env_or_skip,
```

Comment (nfx) [reply tests/integration/aws/test_access.py:114]: this is a nit :)

File: tests/integration/azure/test_locations.py [hunk 22]
```diff
@@ -100,7 +110,11 @@ def test_read_only_location(caplog, ws, sql_backend, inventory_schema):
     azurerm = AzureResources(azure_mgmt_client, graph_client)
 
     location_migration = ExternalLocationsMigration(
-        ws, location_crawler, AzureResourcePermissions(installation, ws, azurerm, location_crawler), azurerm
+        ws,
+        location_crawler,
+        AzureResourcePermissions(installation, ws, azurerm, location_crawler),
```
Comment (nfx) [inline tests/integration/azure/test_locations.py:115]: nit: inline a variable, to be consistent with other context-created objects

File: tests/integration/conftest.py [hunk 23]
```diff
@@ -171,7 +171,7 @@ class TestRuntimeContext(RuntimeContext):  # pylint: disable=too-many-public-met
     def __init__(
         self, make_table_fixture, make_schema_fixture, make_udf_fixture, make_group_fixture, env_or_skip_fixture
     ):
-        super().__init__()
+        RuntimeContext.__init__(self)
```
Comment (nfx) [inline tests/integration/conftest.py:174]: revert

File: tests/integration/conftest.py [hunk 24]
```diff
@@ -201,22 +201,32 @@ def with_dummy_resource_permission(self):
...
+            ]
+            self.with_aws_storage_permissions(instance_profile_mapping, uc_roles_mapping)
 
     def with_azure_storage_permissions(self, mapping: list[StoragePermissionMapping]):
         self.installation.save(mapping, filename=AzureResourcePermissions.FILENAME)
 
-    def with_aws_storage_permissions(self, mapping: list[AWSRoleAction]):
-        self.installation.save(mapping, filename=AWSResourcePermissions.INSTANCE_PROFILES_FILE_NAMES)
+    def with_aws_storage_permissions(
+        self, instance_profile_mapping: list[AWSRoleAction], uc_roles_mapping: list[AWSRoleAction]
```
Comment (nfx) [inline tests/integration/conftest.py:226]:
```suggestion
        self, instance_profile_mapping: list[AWSRoleAction], uc_roles_mapping: list[AWSRoleAction],
```

File: tests/integration/conftest.py [hunk 25]
```diff
@@ -427,10 +448,23 @@ def runtime_ctx(ws, sql_backend, make_table, make_schema, make_udf, make_group,
     return ctx.replace(workspace_client=ws, sql_backend=sql_backend)
 
 
-class LocalAzureCliTest(WorkspaceContext):
-    def __init__(self, _ws: WorkspaceClient, env_or_skip_fixture: Callable[[str], str]):
-        super().__init__(_ws, {})
+class LocalAzureCliTest(TestRuntimeContext, WorkspaceContext):
```
Comment (nfx) [inline tests/integration/conftest.py:451]:
```suggestion
class LocalAzureCliTest(TestRuntimeContext):
```
extending two context factories leads to future hard-to-debug problems. extend only one of the factories - `TestRuntimeContext` in this case.

File: tests/integration/conftest.py [hunk 26]
```diff
@@ -446,8 +480,50 @@ def azure_subscription_id(self):
...
 
 @pytest.fixture
-def az_cli_ctx(ws, env_or_skip):
-    return LocalAzureCliTest(ws, env_or_skip)
+def az_cli_ctx(ws, env_or_skip, make_schema, sql_backend, make_table, make_group, make_udf):
+    ctx = LocalAzureCliTest(ws, env_or_skip, make_schema, make_table, make_udf, make_group)
+    return ctx.replace(workspace_client=ws, sql_backend=sql_backend)
+
+
+class LocalAwsCliTest(TestRuntimeContext, WorkspaceContext):
```
Comment (nfx) [inline tests/integration/conftest.py:488]:
```suggestion
class LocalAwsCliTest(TestRuntimeContext):
```
this hints that `TestRuntimeContext` is not a correct name of the class. It could be renamed to `TestCliContext`.

**Review by nfx — CHANGES_REQUESTED** [review]
contains few hidden bugs
